test/forwarddiff.jl

```diff
@@ -260,3 +260,150 @@ end
     derivative_fd = ForwardDiff.derivative(compute_force, 0.0)
     @test norm(derivative_ε - derivative_fd) < 1e-4
 end
+
+@testitem "Symmetries broken by perturbation are filtered out" tags=[:dont_test_mpi] begin
+    using DFTK
+    using ForwardDiff
+    using LinearAlgebra
+
+    lattice = [2. 0. 0.; 0. 1. 0.; 0. 0. 1.]
+    positions = [[0., 0., 0.], [0.5, 0., 0.]]
+    gauss = ElementGaussian(1.0, 0.5)
+    atoms = [gauss, gauss]
+    atom_groups = [findall(Ref(pot) .== atoms) for pot in Set(atoms)]
+
+    # Select some "interesting" subset of the symmetries
+    # Rotation in the yz plane by 90 degrees
+    rotyz = SymOp([1 0 0; 0 0 1; 0 -1 0], [0., 0., 0.])
+    mirroryz = rotyz * rotyz
+    # Mirror y
+    mirrory = SymOp([1 0 0; 0 -1 0; 0 0 1], [0., 0., 0.])
+    # Translation by 0.5 in the x direction
+    transx = SymOp(diagm([1, 1, 1]), [0.5, 0., 0.])
+
+    is_approx_in(collection, s1) = any(s -> isapprox(s, s1), collection)
+
+    # Generate the full group
+    function complete_group(symops)
+        completed_group = Vector(symops)
+
+        for it = 1:10
+            if it == 10
+                error("Could not complete group in 10 iterations")
+            end
+            to_add = []
+            function check(s1)
+                if !is_approx_in(completed_group, s1) && !is_approx_in(to_add, s1)
+                    push!(to_add, s1)
+                end
+            end
+            # Identity always needs to be there!
+            check(SymOp(diagm([1, 1, 1]), [0., 0., 0.]))
+            for s in completed_group
+                check(inv(s))
+                for t in completed_group
+                    check(s*t)
+                    check(t*s)
+                end
+            end
+            if isempty(to_add)
+                return completed_group
+            end
+            append!(completed_group, to_add)
+        end
+        DFTK.check_group(completed_group)
+    end
+
+    symmetries_full = complete_group([rotyz, mirrory, transx])
+    @test length(symmetries_full) == 16
+
+    DFTK._check_symmetries(symmetries_full, lattice, atom_groups, positions)
+
+    function check_symmetries(filtered_symmetries, expected_symmetries)
+        expected_symmetries = complete_group(expected_symmetries)
+
+        @test length(filtered_symmetries) == length(expected_symmetries)
+        for s in expected_symmetries
+            @test is_approx_in(filtered_symmetries, s)
+        end
+        for s in filtered_symmetries
+            @test is_approx_in(expected_symmetries, s)
+        end
+    end
+
+    # Instantiate Dual to test with perturbations
+    ε = ForwardDiff.Dual{ForwardDiff.Tag{Nothing, Float64}}(0.0, 1.0)
```

Comment:
You could be a little more nasty and use perturbations that are closer to the symmetry tolerance to test the limits.

Reply:
Potentially, but this is not really the goal of this test. Here I just want to ensure that we correctly filter out broken symmetries, not try to play with the heuristic thresholds.